sourcegraph/sourcegraph
# all: final changes for Sourcegraph beta program

- Add a link to `/beta` from GlobalNav (otherwise users can never land on that page,
  as there are no other links to it anywhere / all other uses are inside embedded
  frames).
- Remove the "Email Address" field from the BetaInterestForm; use the account
  primary email instead. This solves a tougher technical aspect of keeping Mailchimp
  in sync, otherwise we would need a bit on each user email address specifying if it
  was the one they registered into Mailchimp or not.
- Add CLI flags to properly manage beta users in all meaningful ways:
  - `src user list [--all-betas=...]`
  - `src user update [--add-betas=...] [--rm-betas=...] [--set-betas=...] <user>`
  - `src user get <user>` now shows betas a user is in.
  - See Quip document below for more details.
- Add wording to the beta page to explain why the form still shows (see BetaInterestForm.js:144-149).
- Tell the user which email address (always their account primary email address) that
  we will be emailing them at.
- When opening the beta form via a modal, always display the success text after
  form submission, and display a close button to exit the modal instead of closing
  it immediately upon submission (which did not show potential errors, either).
- Implement the betas backend, which properly keeps Mailchimp in sync + stores which
  betas a user is part of in Postgres.
- Implement helpers for detecting whether or not a user is in a given beta in both
  Go and JS, see [relevant Quip document describing beta user management](https://sourcegraph.quip.com/KTaTA0pLkt8v) for more details.

Users will need to run the following to migrate to the new DB schema:

```
ALTER TABLE users ADD COLUMN betas text ARRAY;
ALTER TABLE users ADD COLUMN beta_registered boolean;
```

(within the app DB)


##### Reviewer tasks

- I've spent a lot of time making sure this is the _right_ implementation, so please raise any concerns you might have. I've also tested pretty extensively in many different situations.
- Please also briefly review [the Quip document describing beta user management](https://sourcegraph.quip.com/KTaTA0pLkt8v) and comment on anything that stands out.
- I'm very confident it works, but please checkout `sg/beta-cli-server` and run the above DB migration command in `psql`. You can `ALTER TABLE users DROP COLUMN betas;` after to revert the migration. Confirm that nothing weird happens.

##### Test plan

Testing has been quite extensive, but manual. Much of it relies on external services and as such it seems difficult to test under traditional unit tests.

To deploy, I will:
  1. Pin the sourcegraph.com version to master.
  2. Merge this branch to `master`.
  3. Wait for CI build.
  4. Perform the DB migration in prod.
  5. Deploy the new `master` to prod.
  6. Confirm everything looks OK and is working. Any issue can be fixed immediately by rolling back after running the above revertion SQL.
